### PR TITLE
Improve navigation structure

### PR DESCRIPTION
**Pull request by ripcurlx**
Based on https://github.com/bisq-network/bisq/discussions/6092 this is the PR that implements all required changes.
I hope I have covered everything, but it is a lot of testing.

Bildschirmfoto 2022-04-06 um 18 03 16
Bildschirmfoto 2022-04-06 um 18 02 33
Bildschirmfoto 2022-04-06 um 18 02 28
Bildschirmfoto 2022-04-06 um 18 02 24
Bildschirmfoto 2022-04-06 um 18 02 19

Anyone who wants to give it a shot and provide feedback is more than welcome.
This PR also includes improvements in other areas as well (e.g. only Payment methods are shown in filter for selected currency) and many more.

### CODE REVIEW

File: core/src/main/java/bisq/core/locale/CurrencyUtil.java [hunk 1]
```diff
@@ -502,6 +506,114 @@ public static void setBaseCurrencyCode(String baseCurrencyCode) {
         return currencies;
     }
 
+    public static List<TradeCurrency> getAllInteracETransferIdCurrencies() {
```
Comment (ghost) [inline core/src/main/java/bisq/core/locale/CurrencyUtil.java]: What about creating static field with supported currencies in `InteracETransferAccount` instead:
```
//...
public final class InteracETransferAccount extends PaymentAccount {
    public static final List<TradeCurrency> SUPPORTED_CURRENCIES = List.of(new FiatCurrency("CAD"))
//...
``` 
also for other account types (Strike, TikkieId etc.). 

As a alternative add `SUPPORTED_CURRENCIES` into `PaymentMethod` instead of `PaymentAccount`

File: core/src/main/java/bisq/core/payment/PaymentAccountUtil.java [hunk 2]
```diff
@@ -106,6 +113,114 @@ public static boolean isPaymentAccountValidForOffer(Offer offer, PaymentAccount
         return acceptedCountryCodes;
     }
 
+    public static List<TradeCurrency> getTradeCurrencies(PaymentMethod paymentMethod) {
```
Comment (ghost) [inline core/src/main/java/bisq/core/payment/PaymentAccountUtil.java:113]: I think this method breaks [open-closed principle](https://medium.com/@alexandre.malavasi/why-is-the-open-closed-principle-so-important-bed2f2a0d4c7). Refactoring steps suggestion:

1) make `PaymentMethod` abstract class, and create dedicated classes, f.e.:
```
public class SepaPaymentMethod extends PaymentMethod {
    public SepaPaymentMethod () {
        super("SEPA", 6 * DAY, DEFAULT_TRADE_LIMIT_HIGH_RISK);
    }
}
``` 


2) use created classes in constructors of every `PaymentAccount`, f.e.:
```
public SepaAccount() {
    super(new SepaPaymentMethod());
}
```

3. To generate list of supported payment methods use similar approach like in `AssetRegistry`:
```
static {
    for (PaymentAccount account : ServiceLoader.load(PaymentAccount.class)) {
        registeredAccounts.add(account);
    }
}
    
public Stream<PaymentMethod> getAvailablePaymentMethods() {
    return registeredAccounts.stream().map(account -> account.getPaymentMethod());
}
```

* maybe instead loading `PaymentAccount` instances we should load `PaymentMethod` instances - see point 5

4) Add `List<TradeCurrency> getSupportedCurrencies()` into `PaymentMethod`

5) `PaymentMethod` <-> `PaymentAccount` seems to be 1:1 relationship. I think it could be used to additional refactoring. f.e. to replace `PaymentAccountFactory` (which also breaks open-closed principle) with `paymentMethod.createAccountInstance()` 



WDYT ? It could be big refactoring, I didn't analyzed how many changes in codebase it could induce.

**Review by ghost — COMMENTED** [review]
I reviewed only small part of this pull request for now. Let me know what do you think about proposed changes. I'm okay if you disagree with them, however I will be very pleased if they will be introduced :)

File: core/src/main/java/bisq/core/locale/CurrencyUtil.java (hunk 1 above)
Comment (ripcurlx) [reply core/src/main/java/bisq/core/locale/CurrencyUtil.java]: Yes, I agree that it would make more sense to move this information closer to the Domain. I didn't spend too much time yet on bigger refactorings to prevent too many unexpected side effects by changing too much. But yes, I think moving it the PaymentMethod would make sense. Although it is a 1:1 relationship all the time AFAIK, but I think it would be a better fit for the PaymentMethod, which is used by the Payment Account.

Comment (ripcurlx) [reply core/src/main/java/bisq/core/locale/CurrencyUtil.java]: After looking at the code, I revert my opinion and also think adding it to the concrete account implementation is the better way to do this.

File: core/src/main/java/bisq/core/payment/PaymentAccountUtil.java (hunk 2 above)
Comment (ripcurlx) [reply core/src/main/java/bisq/core/payment/PaymentAccountUtil.java:113]: I agree in general with your comment, but this will be a hell of a refactoring and testing. I'd keep it for a separate effort to not completely blow this PR out of proportion.

**Comment by ripcurlx** [comment]
https://user-images.githubusercontent.com/170962/162480084-8393da04-b3d4-4987-84e6-2aa4f922f362.mov


https://user-images.githubusercontent.com/170962/162480111-eaf503a8-551f-4ba4-8c91-96c8aa497fd1.mov


https://user-images.githubusercontent.com/170962/162480114-9b162777-5f88-40d2-91a7-cc78e6295402.mov

**Comment by ghost** [comment]
Checked it out, looks good!  (Some niggles need ironing out).


- Creating BSQ swap offer it somehow is mixing `BSQ Swap` with `Altcoin`, because the offer that gets created has a security deposit Tx and gets displayed like an altcoin offer even though it says "BSQ Swap" as offer type.  When that offer gets taken the trade fails with "hash of payment account is invalid".  This is a clean build from your current branch, run on a fresh regtest env with no altcoin accounts.

- Clicking on an offer in Market View does not always take me to the market containing that offer.  e.g. There is only one offer in the market a Buy BSQ offer.  In Market View click on the offer.  It takes me to Sell BTC for Fiat, and has BSQ selected in the currency drop down.  No offer is shown.

**Comment by ripcurlx** [comment]
> Checked it out, looks good! (Some niggles need ironing out).
> 
> * Creating BSQ swap offer it somehow is mixing `BSQ Swap` with `Altcoin`, because the offer that gets created has a security deposit Tx and gets displayed like an altcoin offer even though it says "BSQ Swap" as offer type.  When that offer gets taken the trade fails with "hash of payment account is invalid".  This is a clean build from your current branch, run on a fresh regtest env with no altcoin accounts.
> * Clicking on an offer in Market View does not always take me to the market containing that offer.  e.g. There is only one offer in the market a Buy BSQ offer.  In Market View click on the offer.  It takes me to Sell BTC for Fiat, and has BSQ selected in the currency drop down.  No offer is shown.

Thanks for testing - I'll look at the BSQ Swap issue you experienced in more detail. It did work for me, so it is kind of weird. I'll post an update today as soon as I'm through with my testing.

@Market View: It can be that there is an issue. I only fixed the behavior on the buy/sell buttons, but not on the triggers in each row. Quite a bit of changes this change caused in the end 😅

**Comment by ripcurlx** [comment]
@jmacxx I just tested it again to create a Buy BSQ and a Sell BSQ offer using BSQ Swaps and it worked as expected for me. Is there a specific flow you used to generate this issue?

The only case where this could happen IMO is that BSQ Swap didn't switch to the BsqSwapCreateOfferView in your case using the regular CreateOfferView.

**Comment by ripcurlx** [comment]
I'll try it with a fresh account without any Altcoin accounts next. Maybe that causes this weird behavior.

**Comment by ripcurlx** [comment]
Ok - now I'm able to reproduce the issue.  It is as I expected caused with a fresh account with only one BSQ account.

**Comment by ripcurlx** [comment]
@jmacxx Just fixed your issue with https://github.com/bisq-network/bisq/pull/6135/commits/1712f7498e58f907929f147c6cdede2610a6b406

**Comment by ripcurlx** [comment]
Adapted the amount and volume description so it matches the fiat case.

**Fiat**
Bildschirmfoto 2022-04-11 um 12 14 33

**Altcoin**
Bildschirmfoto 2022-04-11 um 12 14 15

**Comment by ghost** [comment]
I really like those UI changes! 

A few problems I've found during testing: 

1) It's still possible to add/remove BSQ and XMR on `Settings -> Preferences -> Display altcoins`
2) Got NPE:
```
java.lang.NullPointerException
	at bisq.desktop.main.offer.bisq_v1.MutableOfferViewModel.maybeShowMakeOfferToUnsignedAccountWarning(MutableOfferViewModel.java:1255)
	at bisq.desktop.main.offer.bisq_v1.MutableOfferViewModel.onFocusOutMinAmountTextField(MutableOfferViewModel.java:790)
	at bisq.desktop.main.offer.bisq_v1.MutableOfferViewModel.lambda$onFocusOutAmountTextField$24(MutableOfferViewModel.java:748)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:447)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:446)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Steps to reproduce:
- Go to `Buy -> Other`
- Click `Create new offer to buy`
- Popup `You have not setup a trading account` was displayed (and create offer view in background)
- Click close on popup
- Popup with error appeared


3) Currency combobox is empty on `Buy BSQ -> Create new offer to buy bsq` view:

empty-currency-combobox

Probably it shouldn't be visible there. Trading account could be disabled there, because it's only possible choice.

I don't have BSQ altcoin account setup, only default BSQ swap - it could be tip.

4) Validation errors on  `Buy BSQ -> Create new offer to buy bsq` view are displayed immediately after activation.

**Comment by ripcurlx** [comment]
> 1. It's still possible to add/remove BSQ and XMR on `Settings -> Preferences -> Display altcoins`

True - should be removed there as an option.

> 3. Currency combobox is empty on Buy BSQ -> Create new offer to buy bsq view:

I fixed that incorrectly before. This should be gone now with https://github.com/bisq-network/bisq/pull/6135/commits/2055cc5c104c6371880534d9e5980a9af74b193d

> 4\. Validation errors on  `Buy BSQ -> Create new offer to buy bsq` view are displayed immediately after activation.

Haven't seen that before.

> 2.
I'm pushing a fix soon.

**Comment by ripcurlx** [comment]
Didn't have time until today to address the remaining issues. So here we go.
@xyzmaker123 I didn't remove `XMR` and `BSQ` in the end from the preferences list, as it would have lots of unwanted side effects, so I just display a warning preventing users to remove them from the list.

**Comment by ghost** [comment]
Thanks - looks great! Just two small topics:

1)
> I didn't remove XMR and BSQ in the end from the preferences list, as it would have lots of unwanted side effects

I searched for `preferences.getCryptoCurrencies` in code base and I see only two usages: [A](https://github.com/bisq-network/bisq/blob/fce247cbc0a1f1bf84db9ebbe814b474675a5e79/desktop/src/main/java/bisq/desktop/main/offer/offerbook/OtherOfferBookViewModel.java#L103-L105) and [B](https://github.com/bisq-network/bisq/blob/064335716ad8d34b04f8aff6e21831d09f2a5b81/desktop/src/main/java/bisq/desktop/util/CurrencyList.java#L117). As an alternative approach we could extend this list with XMR and BSQ in B situation, and eliminate `withoutBSQAndTopAltcoin` in filter A. However I don't have strong opinion here - in both cases user could be confused what he is really configuring in preferences.



2) BSQ - altcoin account is visible in combobox (when available), however it's not possible to change to it:

01-bsq-altcoin

**Comment by ghost** [comment]
Issues noticed from mainnet:


- Click Buy XMR.  Sort the list so the best prices are shown at the top.
- Fixed price offers are not sorted correctly.
- Click Sell XMR.  Sort the list so the best prices are shown at the top.
- Fixed price offers are not sorted correctly.

---


- Open Market View for LTC.
- Click on the one offer in the right-hand tableview.
- It took me to the Buy Eth list of open offers.
- Conclusion it navigated to the correct screen but did not refresh the currency to what was chosen.



---


- Scenario:  There is only one LTC offer: maker is selling LTC.
- Click top menu item "Sell", "Other", scroll down the currency list select "LTC Litecoin (1 offer)".
- It takes you to the correct screen, but there are no offers displayed.
- Click top menu item "Buy", "Other", scroll down the currency list select "LTC Litecoin".
- It takes you to the correct screen, but there is 1 offer displayed.
- Conclusion: the dropdown altcoin currency list displays the wrong offer count (market reversed).

**Comment by ripcurlx** [comment]
> * down the currency list select "LTC Litecoin".
> * It takes you to the correct screen, but there is 1 offer displayed.
> * Conclusion: the dropdown altcoin currency list displays the wrong offer count (market reversed).

Thanks for testing - more to fix 😄

**Comment by ghost** [comment]
XMR fixed price offers have the difference from market price negated, resulting in best price sorting not working.


<details><summary>Screenshot compared to v1.8.4</summary>

image


</details>

**Comment by ripcurlx** [comment]
> 

Yes, the sorting is not fixed yet. Was my last open point of the list.

**Comment by ghost** [comment]
I'm testing now, posting issues I found in points below:
1) 
- go to BUY -> Bitcoin
- click "create new offer to buy BTC"
- PROBLEM 1: I see only BSQ swap trading account (I have USD accounts available)
- PROBLEM 2: Label above amount says: "Amount of BTC to spend". I want to buy BTC, so it should says "Amount of BTC to receive"

issue1

**Comment by ripcurlx** [comment]
> * PROBLEM 1: I see only BSQ swap trading account (I have USD accounts available)

hmm that is quite a weird error. It shows the wrong tab selected (it is actually the `Buy BSQ` tab. Ah fck - everytime I fix something for the BSQ swap special handling I break something else 🤦 . Fix is on the way.

**Comment by ghost** [comment]
> Ah fck - everytime I fix something for the BSQ swap special handling I break something else

No problem, it's normal in software developer's life :) OK - I'm waiting for fix to do another iteration of testing.

**Comment by ripcurlx** [comment]
I'll start now with some Mainnet comparisons to find out what else I missed in my Regtest setup.

**Comment by ripcurlx** [comment]
I'm probably already quite blind by all these changes, but I haven't found anything obvious anymore comparing it with v1.8.4 on Mainnet. I'll proceed with the one open refactoring and might do some performance optimizations and clean-ups (and create some new bugs of course 🤪 )

**Comment by ghost** [comment]
Found one "not obvious" kind of issue - preferred currency set to crypto has impact to fiat logic:
- go to Settings -> Preferences
- switch preferred currency to ETH, XMR, BSQ
- go to Buy->Bitcoin or Sell->Bitcoin tab
- see strange behavior (crypto offers visible on list, not possible to create offer)

Sorry for being pain :pleading_face:

**Comment by ripcurlx** [comment]
> Found one "not obvious" kind of issue - preferred currency set to crypto has impact to fiat logic:
> 
> * go to Settings -> Preferences
> * switch preferred currency to ETH, XMR, BSQ
> * go to Buy->Bitcoin or Sell->Bitcoin tab
> * see strange behavior (crypto offers visible on list, not possible to create offer)
> 
> Sorry for being pain 🥺

No worries! Better you find the bug than users. Should be fixed now.

**Comment by ripcurlx** [comment]
After doing some Mainnet tests I found some more edge cases with the preferred currency, which needs some minor improvements - will do it tomorrow.

**Comment by ripcurlx** [comment]
I think it is now in a state ready for a final review testing @jmacxx @xyzmaker123 

There is one UX problem IMO opinion that needs to get solved which we have already for a long time and I'm not 100% sure how we want to solve it best.

Let's assume you want to buy XMR.

Bildschirmfoto 2022-04-21 um 10 17 41

But all those offers doesn't fit your need and you want to create an own offer to buy XMR.

Bildschirmfoto 2022-04-21 um 10 19 58
Bildschirmfoto 2022-04-21 um 10 20 14
Bildschirmfoto 2022-04-21 um 10 20 28
Bildschirmfoto 2022-04-21 um 10 20 33

So what would you expect now? ATM for a new user we close the popup show a notification that you can manage your offers in portfolio. If you close this notification and/or are an existing user it will bring you back to the list of XMR offers that can be bought. But of course there is not your offer visible as it is under Sell > XMR as you created an offer to buy XMR.

Shall we navigate the user instead to the buy XMR offers under Sell > XMR instead?

@UX-P What is your take on this?

**Comment by UX-P** [comment]
@ripcurlx looks good!
The user should land on the Portfolio screen 'My open offers'.  This provides them with immediate access to review and undo their action.

Also, in regards to the create offer screen, instead of 'choose trading account' could the header text read ' Create offer' so that it is clear to the user they are now creating an offer.

**Comment by ghost** [comment]
ACK - I created a few offers on regtest, took them etc. and everything worked okay. Also tested scenarios that has issues before, and now everything is okay. Good job!

**Comment by ripcurlx** [comment]
> Also, in regards to the create offer screen, instead of 'choose trading account' could the header text read ' Create offer' so that it is clear to the user they are now creating an offer.

Sure - I can do the same for "Take offer" as well. I'll adapt the navigational behavior as well.

**Comment by ripcurlx** [comment]
@UX-P It will now look like this:
Bildschirmfoto 2022-04-22 um 10 02 18
Bildschirmfoto 2022-04-22 um 09 53 11

**Comment by ripcurlx** [comment]
The user can now adapt the forth tab by setting a crypto currency as his preferred currency. If no crypto currency is set it will default to `XMR`.

**Example with ETH**
Bildschirmfoto 2022-04-22 um 10 39 07
**XMR as default Altcoin**
Bildschirmfoto 2022-04-22 um 10 39 49

**Review by bisq-github-admin-3 — APPROVED** [review]
ACK - based on https://github.com/bisq-network/bisq/pull/6135#issuecomment-1105152900
